File: src/opponent.py
#!/usr/bin/env python3

import random

import rclpy
import numpy as np
from rclpy.node import Node
from scipy.spatial.transform import Rotation as R

from geometry_msgs.msg import PoseStamped
from ros_gz_interfaces.srv import SetEntityPose

# blue-side waypoints (meters); the opponent is the blue alliance robot
MOVESET = [
    (0.50, 1.50), (1.00, 1.10), (1.55, 0.50), (1.55, -0.50),
    (1.00, -1.10), (0.50, -1.50), (1.30, 0.00), (0.90, 1.55),
    (0.90, -1.55),
]

MOVE_GRAPH = {
    0: [1, 7],
    1: [0, 2, 7],
    2: [1, 3, 6],
    3: [2, 4],
    4: [3, 5, 8],
    5: [4, 8],
    6: [2],
    7: [0, 1],
    8: [4, 5],
}


class Opponent(Node):
    """Simple teleporting opponent robot that wanders the blue side."""

    def __init__(self):
        super().__init__('opponent')

        self.declare_parameter('world_name', 'override')
        self.world_name = self.get_parameter('world_name').value

        self.create_subscription(
            PoseStamped, '/opponent/pose', self.get_opponent_pose, 10)
        self.create_timer(2.0, self.dumb_behavior)

        self.move_pose = self.create_client(
            SetEntityPose, f'/world/{self.world_name}/set_pose')

        self.opp_x = None
        self.opp_y = None
        self.current_pose_key = 0
        self.height = 0.2

    def get_opponent_pose(self, msg: PoseStamped):
        self.opp_x = msg.pose.position.x
        self.opp_y = msg.pose.position.y

    def dumb_behavior(self):
        """wander the blue side pose graph"""
        if np.random.binomial(n=1, p=0.7) == 0:
            return
        potential = MOVE_GRAPH[self.current_pose_key]
        new_index = random.choice(potential)
        self.teleport_to_pose(MOVESET[new_index])
        self.current_pose_key = new_index

    def teleport_to_pose(self, pose):
        req = SetEntityPose.Request()
        req.entity.name = 'opponent'
        req.pose.position.x = pose[0]
        req.pose.position.y = pose[1]
        req.pose.position.z = self.height
        req.pose.orientation.w = 1.0
        self.move_pose.call_async(req)


def main(args=None):
    rclpy.init(args=args)
    node = Opponent()
    try:
        rclpy.spin(node)
    except KeyboardInterrupt:
        pass
    finally:
        node.destroy_node()
        if rclpy.ok():
            rclpy.shutdown()


if __name__ == '__main__':
    main()

File: src/opponent.cpp
// opponent.cpp - a simple teleporting opponent robot that wanders the blue
// side of the Override field via the gz set_pose service.
#include <array>
#include <chrono>
#include <memory>
#include <random>
#include <vector>

#include <geometry_msgs/msg/pose_stamped.hpp>
#include <rclcpp/rclcpp.hpp>
#include <ros_gz_interfaces/srv/set_entity_pose.hpp>

namespace override_sim
{

using SetEntityPose = ros_gz_interfaces::srv::SetEntityPose;

// blue-side waypoints (meters)
const std::vector<std::array<double, 2>> kMoveSet = {
  {0.50, 1.50}, {1.00, 1.10}, {1.55, 0.50}, {1.55, -0.50},
  {1.00, -1.10}, {0.50, -1.50}, {1.30, 0.00}, {0.90, 1.55},
  {0.90, -1.55},
};

const std::vector<std::vector<int>> kMoveGraph = {
  {1, 7}, {0, 2, 7}, {1, 3, 6}, {2, 4}, {3, 5, 8}, {4, 8},
  {2}, {0, 1}, {4, 5},
};

class Opponent : public rclcpp::Node
{
public:
  Opponent() : Node("opponent")
  {
    this->declare_parameter<std::string>("world_name", "override");
    this->get_parameter("world_name", world_name_);

    pose_sub_ = this->create_subscription<geometry_msgs::msg::PoseStamped>(
      "/opponent/pose", 10,
      [this](const geometry_msgs::msg::PoseStamped::SharedPtr msg) {
        opp_x_ = msg->pose.position.x;
        opp_y_ = msg->pose.position.y;
      });
    timer_ = this->create_wall_timer(
      std::chrono::seconds(2), [this]() { dumb_behavior(); });

    set_pose_ = this->create_client<SetEntityPose>(
      "/world/" + world_name_ + "/set_pose");
    height_ = 0.2;
  }

private:
  void dumb_behavior()
  {
    // wander the blue-side pose graph (70% chance each tick)
    if (!bernoulli_(rng_)) {
      return;
    }
    const auto & neighbors = kMoveGraph[current_pose_key_];
    std::uniform_int_distribution<size_t> pick(0, neighbors.size() - 1);
    const int next = neighbors[pick(rng_)];
    teleport_to_pose(kMoveSet[next]);
    current_pose_key_ = next;
  }

  void teleport_to_pose(const std::array<double, 2> & pose)
  {
    if (!set_pose_->service_is_ready()) {
      RCLCPP_WARN(this->get_logger(), "set_pose service not ready");
      return;
    }
    auto req = std::make_shared<SetEntityPose::Request>();
    req->entity.name = "opponent";
    req->pose.position.x = pose[0];
    req->pose.position.y = pose[1];
    req->pose.position.z = height_;
    req->pose.orientation.w = 1.0;
    set_pose_->async_send_request(req);
  }

  std::string world_name_;
  double opp_x_ = 0.0;
  double opp_y_ = 0.0;
  double height_ = 0.2;
  int current_pose_key_ = 0;
  std::mt19937 rng_{std::random_device{}()};
  std::bernoulli_distribution bernoulli_{0.7};

  rclcpp::Subscription<geometry_msgs::msg::PoseStamped>::SharedPtr pose_sub_;
  rclcpp::TimerBase::SharedPtr timer_;
  rclcpp::Client<SetEntityPose>::SharedPtr set_pose_;
};

}  // namespace override_sim

int main(int argc, char ** argv)
{
  rclcpp::init(argc, argv);
  auto node = std::make_shared<override_sim::Opponent>();
  rclcpp::spin(node);
  rclcpp::shutdown();
  return 0;
}
